Objects/rangeobject.c

```diff
@@ -171,6 +171,28 @@ range_dealloc(rangeobject *r)
     PyObject_Free(r);
 }
 
+static unsigned long
+get_len_of_range(long lo, long hi, long step);
+
+/// Return the length as a long, or -1 on error
+static inline long compute_range_length_long(PyObject *start, PyObject *stop, PyObject *step) {
+    int overflow = 0;
+
+    long long_start = PyLong_AsLongAndOverflow(start, &overflow);
+    if (overflow || (long_start==-1 && PyErr_Occurred()) ) {
```

Comment:
Hmm, there's a slightly ugly problem here. If `PyErr_Occurred()` is true then we never clear the exception, so we leave this function with an exception set. That's okay, but then we should be checking `PyErr_Occurred()` in the calling function, too. And yes, I think it's true that given that `start` is a `PyLong_Object`, the _current_ implementation of `PyLong_AsLongAndOverflow` can't possibly raise, so this seems like a non-issue. Except that it's not safe to rely on some _future_ version of `PyLong_AsLongAndOverflow` not raising.

So we either need to check for `PyErr_Occurred()` in the caller, or go back to the idea of a new PyLong API function that's guaranteed not to raise. I think the extra check should be fairly cheap.

Reply:
I think I get the point. The current check `(long_start==-1 && PyErr_Occurred())`  is not good as is never occurs, and if it would occur because the implementation of `PyLong_AsLongAndOverflow` changes we are not handling it correctly. Right?

I refactored so the check is like
```
...
    long long_start = PyLong_AsLongAndOverflow(start, &overflow);
    if (overflow)
        return -1;
    if (long_start==-1 && PyErr_Occurred()) {
        PyErr_Clear();
        return -2;
    }
...
```
For an overflow the documentation of `PyLong_AsLongAndOverflow` states there is no exception. For the other errors we check the value of `long_start` (which is fast) in combination with `PyErr_Occurred()`. If required, we clear the error.

We could also perform the check and clear outside `compute_range_length_long`, but in this way all the logic for the fast path is contained inside a single method.